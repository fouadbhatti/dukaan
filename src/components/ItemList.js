import React from 'react';
import {
  Alert,
  Button,
  ButtonGroup,
  ButtonDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  NavLink
} from 'reactstrap';
import { parse, stringify } from 'query-string';
import PropTypes from 'prop-types';
import { addToCart, setFilters } from '../actions';
import Item from './Item';

class ItemList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      dropdownOpen: false,
      key: 'price'
    };

    this.toggle = this.toggle.bind(this);
    this.onDropdownClick = this.onDropdownClick.bind(this);
    this.onButtonClick = this.onButtonClick.bind(this);
  }

  toggle() {
    this.setState({
      dropdownOpen: !this.state.dropdownOpen
    });
  }

  onDropdownClick(key) {
    const search = parse(location.search);

    search.key = key;
    search.order = this.state.order || 'ASC';

    location.search = stringify(search);
  }

  onButtonClick(val) {
    const search = parse(location.search);

    search.key = this.state.key;
    search.order = val;

    location.search = stringify(search);
  }

  query() {
    const search = parse(location.search);
    const { key, order } = search;

    if (key && order) {
      this.setState({ key, order });
      this.props.dispatch(setFilters({ order, key }));
    }
  }

  componentDidMount() {
    this.query();
  }

  render() {
    const { itemlist, meta, dispatch } = this.props;

    return (
      <div style={{ margin: '90px auto 0 auto' }}>
        <Alert color="light" style={{ textAlign: 'right' }}>
          <ButtonGroup>
            <ButtonDropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
              <DropdownToggle caret outline color="secondary">
                {meta.key}
              </DropdownToggle>
              <DropdownMenu>
                <DropdownItem onClick={() => this.onDropdownClick('price')}>Price</DropdownItem>
                <DropdownItem onClick={() => this.onDropdownClick('name')}>Name</DropdownItem>
              </DropdownMenu>
            </ButtonDropdown>
            <Button
              outline
              color="secondary"
              onClick={() => this.onButtonClick('ASC')}
              active={meta.order === 'ASC'}
              style={{ zIndex: 0 }}
            >
              ↑
            </Button>
            <Button
              outline
              color="secondary"
              onClick={() => this.onButtonClick('DESC')}
              active={meta.order === 'DESC'}
              style={{ zIndex: 0 }}
            >
              ↓
            </Button>
          </ButtonGroup>
        </Alert>
        <div className="row">
          {itemlist.map(item => (
            <div className="col-lg-4 col-md-6  col-xs-6 product-item" key={item.id}>
              <div className="mb-4" style={style}>
                <Item
                  {...item}
                  dispatch={dispatch}
                  children={<Actions dispatch={dispatch} item={item} />}
                />
              </div>
            </div>
          ))}
        </div>
      </div>
    );
  }
}

const style = {
  borderRadius: '3px',
  border: '1px solid #e0e0e0',
  background: '#f7f7f7'
};

ItemList.propTypes = {
  itemlist: PropTypes.array.isRequired,
  meta: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired
};

const Actions = ({ dispatch, item }) => (
  <div style={{ marginTop: '10px' }}>
    <Button
      style={{ marginRight: '10px' }}
      size="sm"
      onClick={() => dispatch(addToCart({ ...item, quantity: 1 }))}
      color="primary"
      className="add-to-cart"
    >
      Add to Cart
    </Button>
    <NavLink href={`/product/${item.id}`} style={{ display: 'inline' }}>
      Details
    </NavLink>
  </div>
);

export default ItemList;
