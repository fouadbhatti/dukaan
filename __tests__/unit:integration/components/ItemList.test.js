import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import ItemList from '../../../src/components/ItemList';

describe('<ItemList />', () => {
  const props = {
    itemlist: [],
    meta: {},
    dispatch: jest.fn()
  };

  beforeEach(() => {});

  it('should render without crashing', () => {
    const wrapper = shallow(<ItemList {...props} />);

    expect(wrapper.length).to.equal(1);
  });

  it('component should have been mounted', () => {
    const spy = sinon.spy(ItemList.prototype, 'componentDidMount');

    const mounted = mount(<ItemList {...props} />);

    expect(spy.callCount).to.equal(1);
  });

  it('should have called click event', () => {
    const wrapper = shallow(<ItemList {...props} />);

    const buttons = wrapper.find('Button');

    expect(buttons).to.have.lengthOf(2);

    const stub = sinon.stub(wrapper.instance(), 'onButtonClick');

    buttons.first().simulate('click');

    expect(stub.callCount).to.equal(1);
  });

  // it('should have a match snapshot', () => {
  //   const wrapper = shallow(<ItemList {...props} />);

  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });
});
