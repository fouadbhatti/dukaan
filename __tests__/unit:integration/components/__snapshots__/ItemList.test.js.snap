// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> should have a match snapshot 1`] = `
<div
  style={
    Object {
      "margin": "90px auto 0 auto",
    }
  }
>
  <Alert
    closeAriaLabel="Close"
    color="light"
    isOpen={true}
    style={
      Object {
        "textAlign": "right",
      }
    }
    tag="div"
    transition={
      Object {
        "appear": true,
        "baseClass": "fade",
        "baseClassActive": "show",
        "enter": true,
        "exit": true,
        "in": true,
        "mountOnEnter": false,
        "onEnter": [Function],
        "onEntered": [Function],
        "onEntering": [Function],
        "onExit": [Function],
        "onExited": [Function],
        "onExiting": [Function],
        "tag": "div",
        "timeout": 150,
        "unmountOnExit": true,
      }
    }
  >
    <ButtonGroup
      role="group"
      tag="div"
    >
      <ButtonDropdown
        isOpen={false}
        toggle={[Function]}
      >
        <DropdownToggle
          aria-haspopup={true}
          caret={true}
          color="secondary"
          outline={true}
        />
        <DropdownMenu
          flip={true}
          tag="div"
        >
          <DropdownItem
            onClick={[Function]}
            tag="button"
            toggle={true}
          >
            Price
          </DropdownItem>
          <DropdownItem
            onClick={[Function]}
            tag="button"
            toggle={true}
          >
            Name
          </DropdownItem>
        </DropdownMenu>
      </ButtonDropdown>
      <Button
        active={false}
        color="secondary"
        onClick={[Function]}
        outline={true}
        style={
          Object {
            "zIndex": 0,
          }
        }
        tag="button"
      >
        ↑
      </Button>
      <Button
        active={false}
        color="secondary"
        onClick={[Function]}
        outline={true}
        style={
          Object {
            "zIndex": 0,
          }
        }
        tag="button"
      >
        ↓
      </Button>
    </ButtonGroup>
  </Alert>
  <div
    className="row"
  />
</div>
`;
